pkg/controllers/dashboardapi/nodedriver/nodedriver.go

```diff
@@ -0,0 +1,61 @@
+package nodedriver
+
+import (
+	"context"
+	"sync/atomic"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/controllers/management/drivers"
+	"github.com/rancher/rancher/pkg/wrangler"
+)
+
+var leader = atomic.Bool{}
+
+func Register(ctx context.Context, wrangler *wrangler.Context) {
+	wrangler.Mgmt.NodeDriver().OnChange(ctx, "custom-node-driver-handler", onChange)
+
+	// when this pod becomes the leader, do not run the onChange code
+	wrangler.OnLeader(func(ctx context.Context) error {
+		leader.Store(true)
+		return nil
+	})
+}
+
+func onChange(_ string, obj *v3.NodeDriver) (*v3.NodeDriver, error) {
+	if obj == nil {
+		return nil, nil
+	}
+
+	// if leader, no need to sync
+	if leader.Load() {
+		return obj, nil
+	}
+
+	if !obj.Spec.Active {
+		return obj, nil
+	}
+
+	// only cache drivers which are not built-in to rancher image
+	if obj.Spec.Builtin {
+		return obj, nil
+	}
+
+	driver := drivers.NewDynamicDriver(obj.Spec.Builtin, obj.Spec.DisplayName, obj.Spec.URL, obj.Spec.Checksum)
+	if driver.Exists() {
+		return obj, nil
+	}
+
+	if err := driver.Stage(true); err != nil {
```

Comment:
Question:  Will the "true" here cause unnecessary redownloading of the same driver?  Should we do some checks like [what this function](https://github.com/rancher/rancher/blob/8e8c62c86d53690a77260884e6fe1afcdafaafc4/pkg/controllers/management/drivers/nodedriver/machine_driver.go#L298-L324) does?

Reply:
This code is what makes it unnecessary: https://github.com/rancher/rancher/blob/c294a41c4253dae7d8de4dfb4ba556134e97b728/pkg/controllers/management/drivers/base_driver.go#L104-L110. The `cacheFile` is in `management-state/machine-drivers/<sha25sum of the URL + driver hash>`, which if it exists means the file is not redownloaded. Now that you mention it, it may be worth gating the `Install` and `Executable` functions behind this `cacheFile` as well, since they don't have similar checks and copy the downloaded binary to the expected path.

Reply:
OK, it sounds like a good enhancement to do.